wfl/generate/minimahopping.py

```diff
@@ -2,100 +2,208 @@
-                     fmax, timestep, totalsteps, skip_failures, **opt_kwargs):
+def _atom_opt_hopping(atom, calculator, Ediff0, T0, minima_threshold, mdmin, parallel, 
+                     fmax, timestep, totalsteps, skip_failures, return_all_traj, maxtemp_scale, **opt_kwargs):
+    fit_idx = opt_kwargs.pop("fit_idx", 0)
+    parallel_seed = opt_kwargs.pop("parallel_seed", None)
+    verbose = opt_kwargs.pop("verbose", True)
+    return_abort_output = opt_kwargs.pop("return_abort_output", False)
+    save_tmpdir = opt_kwargs.pop("save_tmpdir", False)
     workdir = os.getcwd()
-    rundir = tempfile.mkdtemp(dir=workdir, prefix='Opt_hopping_')
+
+#    rundir = tempfile.mkdtemp(dir=workdir, prefix='Opt_hopping_', suffix=str(fit_idx))
+    if save_tmpdir:
+
+        if parallel > 1:
+            rundir = pathlib.Path(f"{workdir}/parallel/{str(parallel_seed).zfill(2)}")
+            if not rundir.exists():
+                rundir.mkdir(parents=True,exist_ok=True)
+        else:	
+            rundir = f"{workdir}/Opt_hopping_{fit_idx}"
+            pathlib.Path(f"{workdir}/Opt_hopping_{fit_idx}").mkdir(parents=True, exist_ok=True)
+    else:
+        rundir = tempfile.mkdtemp(dir=workdir, prefix='Opt_hopping_')
+
     os.chdir(rundir)
     atom.calc = calculator
     try:
         opt = MinimaHopping(atom, Ediff0=Ediff0, T0=T0, minima_threshold=minima_threshold,
                             mdmin=mdmin, fmax=fmax, timestep=timestep, **opt_kwargs)
-        opt(totalsteps=totalsteps)
+        if verbose:
+            print_minhop_parameter(opt)
+        opt(totalsteps=totalsteps, maxtemp=maxtemp_scale*T0)
     except Exception as exc:
+        if parallel_seed is not None and parallel > 1:
+            print(f"parallel run {parallel_seed} failed.")
         # optimization may sometimes fail to converge.
         if skip_failures:
             sys.stderr.write(f'Structure optimization failed with exception \'{exc}\'\n')
             sys.stderr.flush()
             os.chdir(workdir)
-            shutil.rmtree(rundir)
-            return None
+
+            if save_tmpdir:
+                # If this fails at Optimization step, returns optimization trajectory instead. 
+                traj = [] 
+                return _get_after_explosion(rundir, return_abort_output)
+
+            else:
+                shutil.rmtree(rundir)
+                return None
+
         else:
             raise
     else:
         traj = []
-        for hop_traj in Trajectory('minima.traj'):
-            config_type_append(hop_traj, 'hopping_traj')
-            traj.append(hop_traj)
+        if parallel == 1:
+
+            if return_all_traj:
+                traj += _get_MD_trajectory(rundir)
+            for hop_traj in Trajectory('minima.traj'):
+                config_type_append(hop_traj, 'minima')
+                traj.append(hop_traj)
+
+        elif parallel > 1 and parallel_seed is not None:
+            print(f"parallel run {parallel_seed} suceeded.")

```

Comment:
It's a bit tricky in parallel minima hopping. Although each minima hopping is carried out individually, the outcome of whole subprocesses is `minima.traj` which is shared by all different subprocesses. So individual run returns None, but the user can get one `minima.traj` .

Reply:
That's not the way `wfl` works.  If the user wants to take the union of the `minima.traj` that are associated with each of the input configs, that's fine.  But the `wfl` autoparallelization assumption is that the outputs should be mapped one-to-one with the inputs.

The way ASE implements it, writing to a file with a fixed name, is very limiting, especially if you want to parallelize.  I don't see any usefulness in `wfl` replicating that, when it's so easy to do it sensibly, and then the user can take the union of all the outputs if they want. It should run minimahopping for each input config in its own directory, read the files that it returns, and return whatever you think is useful (minima, optimzation trajectories, etc) as the op output.

Reply:
Note that I'm assuming that the combined `minima.traj` is only _written_ to by the various parallel tasks to combine the outputs. If the updated file is also read back in for some reason during the minima hopping process, that's another can of worms, because it's no longer embarrassingly parallel. Relying on things like shared files output will break, e.g., if the inputs are split into multiple remote jobs that may not run on the same machine, or at the same time.

Reply:
Okay, after understanding that, I think maybe parallel minima hopping might not be suitable to `wfl` frame which could be more suitable to high-level features. If this is too complicated to be incorporated into main branch, I'll rather just drop updates in `minimahopping.py`. What do you think?

Reply:
I don't have any strong feelings.  If you think running the separate initial configs entirely separate is not a useful behavior, then we should probably drop it from `wfl`.  However, I want to point out that what you're doing when you just submit entirely separate runs that share a single `minima.traj` file is not reliable.  I looked over the code, and I don't see any evidence that it's written to handle that situation.  What if two processes try to write the new config at the same time? 

If you want multiple searches sharing a single set of minima safely you'd have to write that with explicit MPI parallelization, and while not that complicated, it wouldn't be trivial, because each search can take a different amount of time. One thing you might be able to do is have each search write to its own file, but when they read back the minima they'd read all the files, from all the tasks. That might still have problems if one task tries to read _while_ another is writing, but you can even get around that by writing to a temporary file and renaming it to the normal name once the write is complete.

Reply:
Those potential errors involved in sharing a file between different processes are something I haven't expected. I think implementing MPI parallelized minima hopping is beyond my work scope and there's already another implementation done by Stefan Goedecker's group. Trying this would be quite redundant. So I would rather drop parallel minima from `wfl` and make it rather top-level thing.

Reply:
That's fine by me.  I'll resolve this conversation, and we can keep the remainder (Langevin integrator, mace fitting)